### PR TITLE
Makes Intel 19.0.5 as a default compiler on Compy

### PR DESCRIPTION
**Pull request by singhbalwinder**
[BFB] - Bit-For-Bit

### CODE REVIEW

**Comment by singhbalwinder** [comment]
For some reason I get the following error when I run the model using `SMS.ne4_ne4.FC5AV1C-L` test:
```
40: MPI error (MPI_File_read_at_all) : Other I/O error , error stack:
40: ADIOI_LUSTRE_IOCONTIG(228): Other I/O error No such file or directory
40:  pio_support::pio_die:: myrank=          -1 : ERROR: pionfread_mod.F90.in:
40:          204 : Unknown error occurs in reading file
40: Image              PC                Routine            Line        Source
40: e3sm.exe           00000000028AA6BA  Unknown               Unknown  Unknown
40: e3sm.exe           00000000026863F1  pio_support_mp_pi         120  pio_support.F90
40: e3sm.exe           0000000002683C55  pio_utils_mp_chec          59  pio_utils.F90
40: e3sm.exe           00000000027A262B  pionfread_mod_mp_         204  pionfread_mod.F90.in
40: e3sm.exe           000000000277B9C7  piodarray_mp_read        1064  piodarray.F90.in
40: e3sm.exe           0000000002784B87  piodarray_mp_read         361  piodarray.F90.in
40: e3sm.exe           00000000013AFC97  ncdio_atm_mp_infl         611  ncdio_atm.F90
40: e3sm.exe           000000000169A9F8  inidat_mp_read_in         189  inidat.F90
40: e3sm.exe           00000000013B02BD  startup_initialco          54  startup_initialconds.F90
40: e3sm.exe           00000000012712DF  inital_mp_cam_ini          42  inital.F90
40: e3sm.exe           000000000050BC5B  cam_comp_mp_cam_i         159  cam_comp.F90
40: e3sm.exe           00000000004FFE44  atm_comp_mct_mp_a         312  atm_comp_mct.F90
40: e3sm.exe           000000000043C594  component_mod_mp_         257  component_mod.F90
40: e3sm.exe           000000000042ABFE  cime_comp_mod_mp_        1337  cime_comp_mod.F90
40: e3sm.exe           00000000004394D9  MAIN__                    122  cime_driver.F90
40: e3sm.exe           00000000004188E2  Unknown               Unknown  Unknown
40: libc-2.17.so       00002B2F98CB63D5  __libc_start_main     Unknown  Unknown
40: e3sm.exe           00000000004187E9  Unknown               Unknown  Unknown
```

To get past this error, I have to change `pnetcdf` to `netcdf` in `env_run.xml`.

It dies most likely reading the following file (based on `atm.log.*`) but I am 100% sure:
```
(GETFIL): attempting to find local file USGS-gtopo30_ne4np4_16x.c20160612.nc
 (GETFIL): using
 /compyfs/inputdata/atm/cam/topo/USGS-gtopo30_ne4np4_16x.c20160612.nc
```

Tagging @jayeshkrishna to see if he know what might be going on here.

**Comment by singhbalwinder** [comment]
I haven't tested with PGI but it should work. It had issues due to missing MKLROOT variable and support fixed that. I will give it a try just to be sure.

**Comment by rljacob** [comment]
What netcdf type is that gtopo nc file?

**Comment by singhbalwinder** [comment]
I just tested and it breaks PGI compiler as Intelmpiu5 is not yet built with PGI. I will add a fix for that and push another commit.

I will also find out about the topo.nc file.

**Comment by singhbalwinder** [comment]
```
$ ncdump -k /compyfs/inputdata/atm/cam/topo/USGS-gtopo30_ne4np4_16x.c20160612.nc
64-bit offset
```

**Comment by singhbalwinder** [comment]
Model can now build and run with PGI (I manually added fix from PR #3197). It is interesting that PGI can run fine with PNETCDF unlike Intel. Therefore the problem doesn't seem to be with the topo.nc file....I am still looking into it.

**Comment by singhbalwinder** [comment]
It seems like model compiled with Intel dies while reading "U" variable from file `/compyfs/inputdata/atm/cam/inic/homme/cami_mam3_Linoz_ne4np4_L72_c160909.nc`. Also,
```
$ ncdump -k /compyfs/inputdata/atm/cam/inic/homme/cami_mam3_Linoz_ne4np4_L72_c160909.nc
64-bit offset
```
Everything works fine if I switch `pnetcdf` to `netcdf`.

**Comment by wkliao** [comment]
```
40: MPI error (MPI_File_read_at_all) : Other I/O error , error stack:
40: ADIOI_LUSTRE_IOCONTIG(228): Other I/O error No such file or directory
```
I have seen this error message recently, mentioned by @jedwards4b. (see [lustre bug](https://lists.mcs.anl.gov/pipermail/parallel-netcdf/2019-September/002101.html))
To confirm this is actually a bug in your Intel MPI, I suggest to run the short test program
given in https://github.com/pmodels/mpich/pull/3634. If a similar error message beginning
with "ADIOI_LUSTRE_IOCONTIG(228)” appears, then you got your confirmation (the
remaining error message may be different.)

**Comment by jedwards4b** [comment]
This issue was reported from the TACC system Frontera.
The intel support ticket number is 04349702.

**Comment by singhbalwinder** [comment]
Thanks @wkliao and @jedwards4b . I will stop looking into it further as it is bug in the mpi implementation. 
I do want to note that a slightly older version of the model (with hash d3f4b659e2) does work with this combination (Intel 19.0.5+intelMPI/2019u5). I initially ran that version of the master to confirm that this compiler indeed fixes the issue we saw in the current master (CLUBBv2). Like I mentioned earlier, changing pnetcdf to netcdf also works.

**Comment by wkliao** [comment]
The bug is caused by an uninitialized error code, thus why the 2nd
half message can be different. The error may not occur every time,
depending  on whether the memory content occupied by  that error
code variable at the run time is zero or not.

**Comment by jayeshkrishna** [comment]
Thanks @wkliao and @jedwards4b . It looks like the issue is in the Intel MPI implementation (Intel MPI is derived from MPICH and the fix required in MPICH/ROMIO was https://github.com/pmodels/mpich/commit/3bbdefbf2e3c29b8c42039e08787d028a396ef2b ).

File: cime/config/e3sm/machines/config_machines.xml [hunk 1]
```diff
@@ -1922,21 +1922,24 @@
...
       <modules compiler="pgi">
         <command name="load">pgi/18.10</command>
       </modules>
       <modules mpilib="mvapich2">
         <command name="load">mvapich2/2.3.1</command>
       </modules>
-      <modules mpilib="impi">
-        <command name="load">intelmpi/2018</command>
+      <modules mpilib="impi" compiler="intel">
+	<command name="load">intelmpi/2019u5</command>
```
Comment (jayeshkrishna) [inline cime/config/e3sm/machines/config_machines.xml]: This upgrade might cause issues with MPI I/O like noted in the discussion for this PR

**Comment by singhbalwinder** [comment]
@wkliao : I tried running the program you provided the link for. I have the following in my env:
`module load cmake/3.11.4 intel/19.0.5 intelmpi/2019u5 netcdf/4.6.3 pnetcdf/1.9.0 mkl/2019u5`

I am compiling it using `mpicc` and it compiles fine.

I am running it  using 2 cpus and I get the following error in my error log:
`1: Error: MPI_File_read_at_all()`

It also generates a `testfile.dat` file. I may not be running it correctly. Please let me know.

File: cime/config/e3sm/machines/config_machines.xml (hunk 1 above)
Comment (singhbalwinder) [reply cime/config/e3sm/machines/config_machines.xml]: Yes, I can ask the support to compile an older version of Intel MPI with this new compiler if we know which version introduced this bug.

**Comment by wkliao** [comment]
Hi  @singhbalwinder 
I do not have access to Compy, but believe you  got  it right.
Output file `testfile.dat` is expected  to be generated.
That error message confirms the bug is from Intel MPI library.

**Comment by singhbalwinder** [comment]
Thanks @wkliao. Based on @jayeshkrishna suggestion, I ran @wkliao small program with older MPI implementations (Intel mpi 2019/u3 and 2019/u4) and it seems like this bug shows up only in 2019u5.

@rljacob : Should I ask support to build Intel mpi 2019u4 with intel 19.0.5 to see if it works?

**Comment by jgfouca** [comment]
@-mention me when this is ready to be merged.

**Comment by rljacob** [comment]
@singhbalwinder yes go ahead.

**Comment by drc-pnnl** [comment]
Just to be clear, there is no "building" of Intel MPI. It comes in binary format from Intel.  We can create wrappers for PGI but that's about all that can be done on that front

**Comment by singhbalwinder** [comment]
I have now pushed a commit to change Intel MPI library from 2019u5 to 2019u4. The code runs fine now with the Intel compiler.

**Comment by singhbalwinder** [comment]
If @rljacob approves the additional changes I made after his initial approval, I think this is ready to go. Tagging @jgfouca

**Comment by rljacob** [comment]
Yes those are fine.

**Pull request merged**